### DISCUSSION TITLE
Difficulties in compiling Xyce

### DISCUSSION THREAD

**Discussion by Iachryphagy** [Xyce/Xyce#122] [Q&A] [answered]
I tried to follow the building guide to get Xyce work on my Ubuntu 18.0.4.
Everything seemed fine until a compile problem emerged.
I checked the guide verbatim but can't find any problems.
I have been stuck for days, so I guess I have no choice but to ask here.
Exact procedure below.

Building Trilinos:
```
cmake \
-G "Unix Makefiles" \
-DCMAKE_C_COMPILER="$CC" \
-DCMAKE_CXX_COMPILER="$CPP" \
-DCMAKE_Fortran_COMPILER="$GF" \
-DCMAKE_CXX_FLAGS="$FLAGS" \
-DCMAKE_C_FLAGS="$FLAGS" \
-DCMAKE_Fortran_FLAGS="$FLAGS" \
-DCMAKE_INSTALL_PREFIX=$LIBDIR \
-DCMAKE_MAKE_PROGRAM="make" \
-DTrilinos_ENABLE_NOX=ON \
-DNOX_ENABLE_LOCA=ON \
-DTrilinos_ENABLE_EpetraExt=ON \
 -DEpetraExt_BUILD_BTF=ON \
 -DEpetraExt_BUILD_EXPERIMENTAL=ON \
 -DEpetraExt_BUILD_GRAPH_REORDERINGS=ON \
-DTrilinos_ENABLE_TrilinosCouplings=ON \
-DTrilinos_ENABLE_Ifpack=ON \
-DTrilinos_ENABLE_AztecOO=ON \
-DTrilinos_ENABLE_Belos=ON \
-DTrilinos_ENABLE_Teuchos=ON \
-DTrilinos_ENABLE_COMPLEX_DOUBLE=ON \
-DTrilinos_ENABLE_Amesos=ON \
 -DAmesos_ENABLE_KLU=ON \
-DTrilinos_ENABLE_Amesos2=ON \
 -DAmesos2_ENABLE_KLU2=ON \
 -DAmesos2_ENABLE_Basker=ON \
-DTrilinos_ENABLE_Sacado=ON \
-DTrilinos_ENABLE_Stokhos=ON \
-DTrilinos_ENABLE_Kokkos=ON \
-DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF \
-DTrilinos_ENABLE_CXX11=ON \
-DTPL_ENABLE_AMD=ON \
-DAMD_LIBRARY_DIRS="/usr/lib" \
-DTPL_AMD_INCLUDE_DIRS="/usr/include/suitesparse" \
-DTPL_ENABLE_BLAS=ON \
-DTPL_ENABLE_LAPACK=ON \
$SRCDIR
```
Building Xyce
```
./configure CC="$CC" \
            CXX="$CPP" \
	    F77="$GF"  \
            CXXFLAGS="-O3" \
            ARCHDIR=$LIBDIR \
            --prefix=$INSTALLDIR \
            --enable-stokhos \
	    --enable-amesos2 \
            CPPFLAGS="-I/usr/include/suitesparse" 
```
and upon compiling, a problem always show up
![image](https://github.com/user-attachments/assets/59f8afb4-f63e-4488-8c69-f6964c01d296)
[config.log](https://github.com/user-attachments/files/18275496/config.log)

**Comment by hkthorn** [comment]
Thanks for all the information!  From your config.log and build error, it looks like you are using a version of Trilinos that is newer than 12.12.1.   You can checkout the correct version using the command `git checkout trilinos-release-12-12-branch` in the Trilinos repository.  Then rebuild and reinstall Trilinos before trying to configure/compile Xyce again.  Please completely delete the current Trilinos installation before installing Trilinos 12.12.1 into the `$LIBDIR` directory.  Otherwise newer and older Trilinos artifacts will exist in the same directory and may cause other configure/compile errors with Xyce.

**Reply by tvrusso** [reply]
Heh.  I must have been typing as you were writing this.  In fact, as we found with #121, this *exact* error is caused by having bits and pieces of two different versions of Trilinos lying around  (that is, the presence of artifacts of dissimilar versions doesn't just cause "other configure/compile errors" --- it directly causes *this* compile error).

**Comment by tvrusso** [comment, marked as answer]
This is pretty much an exact duplicate of #121, which was just solved right around Christmas and only closed the day before yesterday.

The problem there was that the user had accidentally had left bits and pieces of a new version of Trilinos installed in a place that configure could find alongside a second install of Trilinos version 12.12.1.  I suspect you've got the same problem.

The issue is that the file `NOX_SolverStats.hpp` is being found, which causes configure to think that you're using a new version of Trilinos, but all other files in the Trilinos install are from an older version (the recommended version 12.12.1).  The result is that Xyce declares the `getStatus` method of the `PseudoTransientBased` class differently from the way that the base class is defined in Trilinos 12.12.1 --- Trilinos changed the declaration of that base class virtual function, and if the definition in Xyce doesn't match the declaration in Trilinos, Xyce's function doesn't override the base class and the base class remains pure virtual.

This can happen multiple ways.  

Did you by any chance first install a recent version of Trilinos, then  go back and reinstall 12.12.1?   If you didn't clear out the target directory, then the new install left traces of the old install, because newer Trilinos has some files that older Trilinos versions didn't have --- those wouldn't get deleted or rewritten, just left lying around.  One of those is the `NOX_SolverStats.hpp` file that is causing this problem.

There could be other ways this might happen (the person who opened #121 had installed a new Trilinos directly into /usr/include and /usr/lib, and then tried to remove it (but didn't do so completely), and then installed version 12.12.1 into a completely local directory in their home directory --- as a result, Xyce's configure found the left over `NOX_SolverStats.hpp` from the /usr/include directory, but all the other files it needed from the second install.

It is not clear where you are installing Trilinos since your cmake invocation references `$LIBDIR` but you haven't shown what that variable is set to.

Can you check to see if the file `NOX_SolverStats.hpp` exists on your system?   `find / -name NOX_SolverStats.hpp` should pin it down.   Once that's found, you'll know where the wrong version of Trilinos is hiding and can clean it up.

**Reply by tvrusso** [reply]
On looking at your config.log it is now clear you've installed Trilinos into `/home/trap/Desktop/xyce_old_build/xyce/xyce_lib/` so my "find" command is overkill.  I bet you will find that `/home/trap/Desktop/xyce_old_build/xyce/xyce_lib/include` has `NOX_SolverStats.hpp` in it, and if it does it's because you tried installing a newer version of Trilinos there first and didn't completely delete it.  

Clean out your ARCHDIR completely and reinstall Trilinos 12.12.1 and your problems should go away.

If that directory does NOT have `NOX_SolverStats.hpp` then it must exist somewhere else, because your configure run is definitely  finding it.  In that case, you'd have to try my "find" command above to locate and eradicate it (and the rest of the wrong version of Trilinos with it).

**Reply by Iachryphagy** [reply]
Thank you for your quick answer!
That makes sense, because I tried to follow the INSTALL.md in the repository, and it says you should download a Trilinos of version 14.4, so did I, but I failed to finish the whole process, (that is why the path I'm working on is named "old_xyce"), then I turned to the online building guide.
I will clean the leftovers up, and try again, if it works, I will leave a message here.

**Comment by Iachryphagy** [comment]
Thanks folks, I finally solved this problem, Mr.Russo has made the reasons very clear.
It turns out that, like, the compiling procedue was effected by some pieces of Trilinos remain in its default path or somewhere else. I believe I had messed the authority level so some parts of trilinos failed to be removed by the unistall program,
I tried my script on another dedicated mechine, and this time it works just fine.
Thank you again for your quick response!
